src/commands/cmd_query.c

```diff
@@ -65,43 +61,45 @@ void _MGraph_Query(void *args) {
     RedisModuleCtx *ctx = RedisModule_GetThreadSafeContext(qctx->bc);
     AST_Query* ast = qctx->ast;
     ResultSet* resultSet = NULL;
-    const char *graph_name = RedisModule_StringPtrLen(qctx->graphName, NULL);
+
+    bool readonly = AST_ReadOnly(ast);
+    // If the query modifies the keyspace, acquire the Redis global lock
+    // TODO I think this locking may be overzealous - revisit later
+    // and see if we can reduce the lock's lifetime
+    if (!readonly) RedisModule_ThreadSafeContextLock(ctx);
```

Comment:
Am I wrong about this? The documentation seems to indicate that we should wrap sensitive API calls in a thread-safe context lock. If we do it like this, I don't think having read-write locks on separate GraphContexts will have any effect; since any writer will still lock the Redis GIL.